website/blog/2023-01-17-grouping-data-tests.md

```diff
@@ -0,0 +1,96 @@
+---
+title: "Make grouping a first-class citizen in your data quality checks"
+description: "Which of these numbers doesn't belong? [-1, 0, 1, null]. You can't judge data quality without data context, so our tools should enable as much context as possible."
+slug: grouping-data-tests
+
+authors: [emily_riederer]
+
+tags: [analytics craft]
+hide_table_of_contents: false
+
+date: 2023-01-17
+is_featured: true
+---
+Imagine you were responsible for monitoring the safety of a subway system. Where would you begin? Most likely, you'd start by thinking about the key risks like collision or derailment, contemplate what causal factors like scheduling software and track conditions might contribute to bad outcomes, and institute processes and metrics to detect if those situations arose. What you wouldn't do is blindly apply irrelevant industry standards like testing for problems with the landing gear (great for planes, irrelevant for trains) or obsessively worry about low probability events like accidental teleportation before you'd locked down the fundamentals. 
+
+When thinking about real-world scenarios, we're naturally inclined to think about key risks and mechanistic causes. However, in the more abstract world of data, many of our data tests often gravitate towards one of two extremes: applying rote out-of-the-box tests (nulls, PK-FK relationships, etc.) from the world of traditional database management or playing with exciting new toys that promise to catch our wildest errors with anomaly detection and artificial intelligence. 
+
+Between these two extremes lies a gap where human intelligence goes. As with real-world systems, analytics engineers could create more effective tests by embedding the context they have on the real-world data _generation_ process and the technical data _production_ process (a subject which [I've written about previously](https://emilyriederer.com/post/data-error-gen/)). While such expressive tests will be unique to our domain, modest tweaks to our mindset can help us implement them with our standard tools. This post demonstrates how the simple act of conducting tests _by group_ can expand the universe of possible tests, boost the sensitivity of the existing suite, and help keep our data "on track". This feature is [now available in dbt-utils](https://github.com/dbt-labs/dbt-utils#grouping-in-tests). 
```

Comment:
I think it would be useful to clarify the difference between generation and production a bit (or is that what your linked article is meant to cover? I didn't see a real disambiguation component in that article when I searched)

Reply:
I think what I meant by that is just:

we have a **data generating process** of things that happen in the real world. These can be accurately reflected in our data but inconvenient. Like, after turnstile replacement, it is objectively _true_ that the turnstile entries drops to zero because that's literally what happened. However, it's inconvenient

we then have a **data production process** of things built into our ETL process that can cause error 'artifacts' that are neither true nor convenient. For example, once I take a lagged different of the turnstile data that dropped to zero, my data might say I had negative entries for a given period, which is just wrong 

High-level, this relates to that other blog post as the first is more (real world / data collection) and the second is more (data loading / data transformation). That said, it's possible this is all far too much of a rabbit hole to go down. The statistician crowd talks quite a lot about the data generating process for modeling purposes, so it's more of a helpful analogy there than writing for an AE audience, I think

What do you think about replacing the paragraph with this?

```
Between these two extremes lies a gap where human intelligence goes. As with real-world systems, analytics engineers could create more effective tests by embedding the context they have on the reality that the data represents and the liabilities created by its transformations (many of which I've catalogued [here](https://emilyriederer.com/post/data-error-gen/)). While such expressive tests will be unique to our domain, modest tweaks to our mindset can help us implement them with our standard tools. This post demonstrates how the simple act of conducting tests _by group_ can expand the universe of possible tests, boost the sensitivity of the existing suite, and help keep our data "on track". This feature is [now available in dbt-utils](https://github.com/dbt-labs/dbt-utils#grouping-in-tests). 
```

I'm also good with taking out the link to the other blog altogether. Your call!

Reply:
I reshuffled it a bit to 

>Analytics engineers can create more effective tests by embedding their understanding of how the data was created, and especially how this data can go awry (a topic I've [written about previously](https://emilyriederer.com/post/data-error-gen/))

what do you think?

Reply:
Sounds good to me - thank you!